### PR TITLE
Operator on Fabric8 actions

### PR DESCRIPTION
**Pull request by andreaTP**
Opening a Draft PR to monitor the progress and to share.

Current status:
 - tests are passing
 - IT tests are passing
 - all the follow-up PRs (helm / release / enterprise operator) have been drafted

TODO:
 - @RayRoestenburg review
 - self-review
 - more testing on live applications
 - fix the headers

### CODE REVIEW

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/Operator.scala [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+  }
+
+  val AppIdLabel = "com.lightbend.cloudflow/app-id"
+  val ConfigFormatLabel = "com.lightbend.cloudflow/config-format"
+  val StreamletNameLabel = "com.lightbend.cloudflow/streamlet-name"
+  val ConfigUpdateLabel = "com.lightbend.cloudflow/config-update"
+
+  val DefaultWatchOptions = Map(CloudflowLabels.ManagedBy -> CloudflowLabels.ManagedByCloudflow)
+
+  val MaxObjectBufSize = 8 * 1024 * 1024
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/Operator.scala:71]: This is likely not used anymore, can be removed?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/Operator.scala [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+  val restartSettings = RestartSettings(minBackoff = 3.seconds, maxBackoff = 30.seconds, randomFactor = 0.2)
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/Operator.scala:73]: Can likely be removed?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/Operator.scala [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+  val StreamAttributes = ActorAttributes.supervisionStrategy(decider)
+
+  private lazy val fabric8ExecutionContext = ExecutionContext.fromExecutorService(Executors.newSingleThreadExecutor())
+
+  def handleEvents(client: KubernetesClient, runners: Map[String, Runner[_]], podName: String, podNamespace: String)(
+      implicit system: ActorSystem,
+      mat: Materializer,
+      ec: ExecutionContext) = {
+
+    // handleAppEvents
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/Operator.scala:88]: why not have separate methods?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/Operator.scala [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+    source
+  }
+
+  private val podInformer = new AtomicReference[SharedIndexInformer[Pod]]()
+
+  private def watchPod(sharedInformerFactory: SharedInformerFactory, options: Map[String, String])(
+      implicit system: ActorSystem): Source[WatchEvent[Pod], NotUsed] = {
+
+    // TODO probably should be an atomic ref as well
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/Operator.scala:258]: Remove comment?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/ActionExtension.scala [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import io.fabric8.kubernetes.api.model.{ Secret, SecretList }
+import io.fabric8.kubernetes.client.KubernetesClient
+import io.fabric8.kubernetes.client.dsl.{ MixedOperation, Resource }
+
+import scala.util.{ Failure, Success, Try }
+
+object ActionExtension {
+
+  // This is needed since, sometimes, the secrets created by the CLI are taking time to be materialized
+  def providedRetry(name: String, namespace: String)(fAction: Option[Secret] => Action)(
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/ActionExtension.scala:29]:
```suggestion
  def providedSecretRetry(name: String, namespace: String)(fAction: Option[Secret] => Action)(
```

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Actions.scala [hunk 6]
```diff
@@ -34,59 +37,45 @@ object Actions {
...
       podName: String,
       podNamespace: String,
-      cause: ObjectResource
-  ): Seq[Action] = {
+      cause: ObjectReference): Seq[Action] = {
     require(currentApp.forall(_.spec.appId == newApp.spec.appId))
-    val labels          = CloudflowLabels(newApp)
-    val ownerReferences = CloudflowApplication.getOwnerReferences(newApp)
+    val labels = CloudflowLabels(newApp)
+    val ownerReferences = List(Util.getOwnerReference(newApp.name, newApp.getMetadata.getUid))
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Actions.scala:48]: Rather create an object with apply called AppOwnerReference, so this becomes:
```
val ownerReferences = List(AppOwnerReference(newApp.name, newApp.getMetadata.getUid))
```
`Util` class pet peeve triggered ;-)

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/CloudflowStatus.scala [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    oldApp.map(_.getStatus) match {
+      case Some(status) if status.appStatus == Some(Status.Error) => false
+      case _                                                      => true
+    }
+  }
+
+  implicit val adapter =
+    CustomResourceAdapter[App.Cr, App.List](App.customResourceDefinitionContext)
+
+  // TODO: this failure still happen, seldom and not reproduciple ...
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/CloudflowStatus.scala:257]:
```suggestion
  // TODO: this failure still happen, seldom and not reproducible ...
```

Comment (RayRoestenburg) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/CloudflowStatus.scala:257]: What is the failure @andreaTP ?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/CloudflowStatus.scala [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+                App.Cr(spec = null, metadata = null)
+              }
+          }
+        }
+      }, { res =>
+        res match {
+          case Success(_) => Action.noop
+          case Failure(err) if retry > 0 =>
+            log.error(s"Failure updating the CR status retries: $retry", err)
+            Thread.sleep(100)
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/CloudflowStatus.scala:290]: ehhhh

Comment (RayRoestenburg) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/CloudflowStatus.scala:290]: lol

Comment (RayRoestenburg) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/CloudflowStatus.scala:290]: Retry 5 times 100ms in between? Is this some 'grease' for tests that very quickly undeploy/deploy the same CR?

Comment (RayRoestenburg) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/CloudflowStatus.scala:290]: Maybe worth to discuss this one separately.

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Util.scala [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+      .withKind(App.Kind)
+      .withName(name)
+      .withUid(uid)
+      .build()
+  }
+
+  // TODO: we should keep a single source of truth for model
+  import cloudflow.blueprint.VolumeMountDescriptor
+  import cloudflow.blueprint.deployment._
+  def toBlueprint(deployment: App.Deployment): StreamletDeployment = {
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Util.scala:30]: Why is this not a method on `App.Deployment`?

Comment (RayRoestenburg) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Util.scala:30]: Or in `StreamletDeployment` object, as `fromBlueprint`?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Defaults.scala [hunk 10]
```diff
@@ -26,50 +30,49 @@ sealed trait RunnerDefaults {
...
 final case class FlinkRunnerDefaults(
     parallelism: Int,
     jobManagerDefaults: FlinkJobManagerDefaults,
     taskManagerDefaults: FlinkTaskManagerDefaults,
-    prometheusRules: String
-) extends RunnerDefaults
+    prometheusRules: String)
+    extends RunnerDefaults
+
+object Common {
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Defaults.scala:72]: Change to `JsonConfig` object with apply that returns `Config`, or just copy paste that one line in the two places you use it.

Also does not really seem to fit in `Defaults` file...

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EndpointActions.scala [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+  private def serviceResource(
+      endpoint: App.Endpoint,
+      streamletDeploymentName: String,
+      namespace: String,
+      labels: CloudflowLabels,
+      ownerReferences: List[OwnerReference]): Service = {
+    val servicePort = {
+      new ServicePortBuilder()
+      // TODO: Check if 3000 is a good default
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EndpointActions.scala:69]: Probably outdated doc.

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EndpointActions.scala [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        .withName(Name.ofContainerPort(endpoint.containerPort.getOrElse(3000)))
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EndpointActions.scala:70]: very minor but put 3000 in a val

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EndpointActions.scala [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+          .withPorts(servicePort)
+          .withSelector(Map(CloudflowLabels.Name -> Name.ofPod(streamletDeploymentName)).asJava)
+          .build())
+      .build()
+  }
+
+  private def createServiceAction(endpoint: App.Endpoint, app: App.Cr, streamletDeploymentName: String): Action = {
+    val labels = CloudflowLabels(app)
+    val ownerReferences = List(
+      // TODO: pretty sure this is repeated somewhere
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EndpointActions.scala:98]: This is likely the `AppOwnerReference`

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EventActions.scala [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+import akka.kube.actions.Action
+import cloudflow.operator.action.EventActions.EventType.EventType
+import cloudflow.operator.action.runner.Runner
+import io.fabric8.kubernetes.api.model.{ EventBuilder, EventSourceBuilder, ObjectReference }
+
+import java.time.ZonedDateTime
+import java.time.format.DateTimeFormatter
+import scala.jdk.CollectionConverters._
+
+object EventActions {
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EventActions.scala:28]: Note to self: Continue here...

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/CloudflowStatus.scala (hunk 7 above)
Comment (andreaTP) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/CloudflowStatus.scala:257]: fail to update the Status

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/Operator.scala (hunk 3 above)
Comment (andreaTP) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/Operator.scala:88]: to share the `sharedInformerFactory` across the streams, we can split it up, but, in case we should push to the `Main` the `informerFactory`, as you prefer ...

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Actions.scala (hunk 6 above)
Comment (andreaTP) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Actions.scala:48]: I understand, it was a stub 👍  removed

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Util.scala (hunk 9 above)
Comment (andreaTP) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Util.scala:30]: moved as a private method to the `Runner`

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Defaults.scala (hunk 10 above)
Comment (andreaTP) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/Defaults.scala:72]: I like `JsonConfig`

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EndpointActions.scala [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+    val getOperation = { client: KubernetesClient =>
+      (client.services().asInstanceOf[MixedOperation[Service, ServiceList, Resource[Service]]])
+    }
+
+    val executeOperation = { services: MixedOperation[Service, ServiceList, Resource[Service]] =>
+      val thisService =
+        services.inNamespace(service.getMetadata.getNamespace).withName(service.getMetadata.getName)
+      val ser = thisService.get()
+      Option(ser) match {
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EndpointActions.scala:118]: Maybe do `Option(thisService.get())` so no one can make the mistake to use ser (super minor)

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EventActions.scala [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+      reason: String,
+      message: String,
+      `type`: EventType = EventType.Normal,
+      objectReference: ObjectReference,
+      fieldPath: Option[String] = None): Action = {
+    val eventTime = ZonedDateTime.now()
+    val metadataName = newEventName(podName, app.spec.appId)
+
+    // the object reference fieldPath is irrelevant for application events.
+    fieldPath match {
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EventActions.scala:114]: minor, use `foreach`?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/PrepareNamespaceActions.scala [hunk 17]
```diff
@@ -16,21 +16,25 @@
...
-            labels: CloudflowLabels,
-            ownerReferences: List[OwnerReference]): Seq[Action] =
+  def apply(
+      app: App.Cr,
+      runners: Map[String, Runner[_]],
+      labels: CloudflowLabels,
+      ownerReferences: List[OwnerReference]): Seq[Action] =
     app.spec.streamlets
-      .map(streamlet => streamlet.descriptor.runtime.name)
+      .map(streamlet => streamlet.descriptor.runtime)
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/PrepareNamespaceActions.scala:37]: Why was this change needed?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+import scala.jdk.CollectionConverters._
+
+/**
+ * Creates topic actions for managed topics.
+ */
+object TopicActions {
+
+  private val log = LoggerFactory.getLogger(TopicActions.getClass)
+
+  // TOOD: again Blueprint and CR overlapping
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala:45]:
```suggestion
  // TODO: again Blueprint and CR overlapping
```

Comment (RayRoestenburg) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/TopicActions.scala:45]: I think we know what to do about this, so maybe remove the comment.

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala [hunk 19]
```diff
@@ -0,0 +1,465 @@
...
+/**
+ * Creates the Resources that define an Akka [[Runner]].
+ */
+final class AkkaRunner(akkaRunnerDefaults: AkkaRunnerDefaults) extends Runner[Deployment] {
+  import AkkaRunner._
+  import akkaRunnerDefaults._
+
+  val runtime = Runtime
+
+  def createOrReplaceResource(res: Deployment)(implicit ct: ClassTag[Deployment]): Action = {
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:54]: not used?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala [hunk 20]
```diff
@@ -0,0 +1,465 @@
...
+    Action.createOrReplace(res)
+  }
+
+  def deleteResource(name: String, namespace: String)(implicit ct: ClassTag[Deployment]): Action =
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:58]: not used?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala [hunk 21]
```diff
@@ -0,0 +1,465 @@
...
+
+    def execute(client: KubernetesClient)(implicit ec: ExecutionContext): Future[Action] = {
+      Future {
+        client
+          .apps()
+          .deployments()
+          .inNamespace(deployment.getMetadata.getNamespace)
+          .withName(deployment.getMetadata.getName)
+          .patch(deployment)
+        Action.log.warn("Akka deployment patched!")
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:81]: Why warn? maybe debug leftover.

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala [hunk 22]
```diff
@@ -0,0 +1,465 @@
...
+        Action.noop
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:82]: Why return noop? you could return `this`?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala [hunk 23]
```diff
@@ -0,0 +1,465 @@
...
+  }
+
+  def prometheusConfig = PrometheusConfig(prometheusRules)
+
+  def resource(
+      deployment: App.Deployment,
+      app: App.Cr,
+      configSecret: Secret,
+      updateLabels: Map[String, String] = Map()): Deployment = {
+    // The runtimeConfig is already applied in the runner config secret, so it can be safely ignored.
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:166]: Which runtimeConfig? maybe outdated comment, remove?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala [hunk 24]
```diff
@@ -0,0 +1,465 @@
...
+        .withName(Name.ofContainerPrometheusExporterPort)
+        .build()
+
+    val podsConfig = getPodsConfig(configSecret)
+
+    // TODO check if this is still valid.
+    // Pass this argument to the entry point script. The top level entry point will be a
+    // cloudflow-entrypoint.sh which will route to the appropriate entry point based on the
+    // arguments passed to it
+    val args = List("akka")
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:200]: I think this can be removed, the entrypoint is now already akka specific, so please remove comment and these args.

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala [hunk 25]
```diff
@@ -0,0 +1,465 @@
...
+    val resourceRequirements = createResourceRequirements(podsConfig)
+    val environmentVariables = createEnvironmentVariables(app, podsConfig)
+
+    val c: ContainerBuilder = {
+      new ContainerBuilder()
+        .withName(podName)
+        .withResources(resourceRequirements)
+        .withImage(deployment.image)
+        .withEnv(environmentVariables: _*)
+        .withArgs(args: _*)
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:279]: remove this

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala [hunk 26]
```diff
@@ -0,0 +1,465 @@
...
+    val limits = {
+      resourceConstraints.cpuLimits.map(v => "cpu" -> v).toMap ++
+      resourceConstraints.memoryLimits.map(v => "memory" -> v).toMap
+    }
+
+    val requests = {
+      Map("cpu" -> resourceConstraints.cpuRequests, "memory" -> resourceConstraints.memoryRequests)
+    }
+
+    lazy val resourceRequirements =
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:410]:
```suggestion
    lazy val resourceRequirementsFromSettings =
```

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala [hunk 27]
```diff
@@ -0,0 +1,465 @@
...
+        .build()
+
+    // TODO slightly changed the logic / doublecheck the correctness
+    // e.g. now we go with "less defaults" probably
+    (for {
+      pod <- podsConfig.pods.get(PodsConfig.CloudflowPodName)
+      containerConfig <- pod.containers.get(PodsConfig.CloudflowContainerName)
+      resources <- containerConfig.resources
+    } yield {
+      resources
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:423]: you miss the defaults here.

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/FlinkRunner.scala [hunk 28]
```diff
@@ -0,0 +1,580 @@
...
+
+  private def getResourceRequirements(
+      resourceDefaults: FlinkPodResourceDefaults,
+      podsConfig: PodsConfig,
+      podName: String): Option[ResourceRequirements] = {
+
+    var resReqManagerBuilder = new ResourceRequirementsBuilder()
+
+    val defaultRequests =
+      ((resourceDefaults.cpuRequest match {
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/FlinkRunner.scala:255]: map?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/FlinkRunner.scala [hunk 29]
```diff
@@ -0,0 +1,580 @@
...
+        case Some(req) => Map("cpu" -> req)
+        case _         => Map.empty[String, Quantity]
+      }) ++
+      (resourceDefaults.memoryRequest match {
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/FlinkRunner.scala:259]: map?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/Runner.scala [hunk 30]
```diff
@@ -0,0 +1,537 @@
...
+    val createActions = newDeployments
+      .filterNot(deployment => currentDeploymentNames.contains(deployment.name))
+      .flatMap { deployment =>
+        Seq(
+          Action.createOrReplace(configResource(deployment, newApp)),
+          ActionExtension.providedSecretRetry(deployment.secretName, newApp.namespace)({
+            case Some(secret) =>
+              createOrReplaceResource(resource(deployment, newApp, secret))
+            case None =>
+              // TODO: this is an un-recoverable error -> errorAction instead? @Ray
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/Runner.scala:141]: Let's not change this now.

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/Runner.scala [hunk 31]
```diff
@@ -0,0 +1,537 @@
...
+
+  private def getContainerConfig(container: CloudflowConfig.Container): ContainerConfig = {
+    val env = container.env.map { env =>
+      new EnvVarBuilder()
+        .withName(env.name)
+        .withValue(env.value)
+        .build()
+    }
+
+    val resources = {
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/Runner.scala:433]: 👍

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/SparkRunner.scala [hunk 32]
```diff
@@ -0,0 +1,654 @@
...
+
+    log.debug(s"""
+    Streamlet ${deployment.streamletName} - resources for executor pod:
+      coreRequest:    ${updatedExecutor.coreRequest}
+      coreLimit:      ${updatedExecutor.coreLimit}
+    """)
+    updatedExecutor
+  }
+
+  // TODO: changed the original logic that seem to be incorrect
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/SparkRunner.scala:375]: We discussed, can remove comment.

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/SparkRunner.scala [hunk 33]
```diff
@@ -0,0 +1,654 @@
...
+      Future {
+        val typedSelector = client
+          .customResources(customResourceDefinitionContext, classOf[Cr], classOf[List])
+          .inNamespace(cr.namespace)
+          .withName(cr.name)
+
+        Option(typedSelector.fromServer().get()) match {
+          case Some(curr) =>
+            // NOTE: the typed API for patching doesn't work ...
+            Action.log.warn(s"Patching Spark Cr ${cr.name} in ${cr.namespace}")
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/SparkRunner.scala:611]: warn?

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/event/StreamletChangeEvent.scala [hunk 34]
```diff
@@ -0,0 +1,138 @@
...
+            if (configFormat == CloudflowLabels.StreamletDeploymentConfigFormat) {
+              log.debug(s"[app: ${streamletChangeEvent.appId} updating config for streamlet $streamletName]")
+              actionsForRunner(app, streamletChangeEvent, podName, runners)
+            } else Nil
+          }
+          .getOrElse(Nil)
+      case _ => Nil // app could not be found, do nothing.
+    }
+
+  // TODO: duplicated fixme
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/main/scala/cloudflow/operator/event/StreamletChangeEvent.scala:104]: ?

File: tools/cloudflow-operator/src/test/scala/cloudflow/operator/action/CloudflowApplicationSpec.scala [hunk 35]
```diff
@@ -0,0 +1,376 @@
...
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import io.fabric8.kubernetes.api.model._
+import io.fabric8.kubernetes.client.utils.Serialization
+import org.scalatest.matchers.must.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+import org.scalatest.{ EitherValues, GivenWhenThen, Inspectors, OptionValues }
+
+import scala.jdk.CollectionConverters._
+
+class CloudflowApplicationSpec
```
Comment (RayRoestenburg) [inline tools/cloudflow-operator/src/test/scala/cloudflow/operator/action/CloudflowApplicationSpec.scala:32]: AppCrSpec?

File: tools/project/Dependencies.scala [hunk 36]
```diff
@@ -68,4 +88,27 @@ object Dependencies {
   val cloudflowNewItLibrary =
     libraryDependencies ++= Seq(Compile.commonsCodec, Compile.commonsCompress, Compile.scalatest)
 
+  val cloudflowBlueprint =
+    libraryDependencies ++= Seq(
+        Compile.typesafeConfig,
+        Compile.sprayJson,
+        // TODO: check if Avro and ScalaPB can stay in a separate module
```
Comment (RayRoestenburg) [inline tools/project/Dependencies.scala:95]: 👍

**Review by RayRoestenburg — APPROVED** [review]
LGTM in general, some minor comments, and the fix for pod config resource requirements.

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/PrepareNamespaceActions.scala (hunk 17 above)
Comment (andreaTP) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/PrepareNamespaceActions.scala:37]: it's related to the usage of the Fabric8 CR

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala (hunk 19 above)
Comment (andreaTP) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:54]: used in the Runner

File: tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala (hunk 20 above)
Comment (andreaTP) [reply tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/runner/AkkaRunner.scala:58]: in the Runner
